pkg/admin/migration/migration.go

```diff
@@ -0,0 +1,291 @@
+package migration
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	crdclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	v1 "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/retry"
+	apiserviceclient "k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset"
+	apiregistrationv1 "k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset/typed/apiregistration/v1"
+
+	"k8s.io/klog/v2"
+)
+
+var excludeResources = sets.NewString(
+	"events",
+)
+
+var metadataAccessor = meta.NewAccessor()
+
+type migrator struct {
+	client           dynamic.Interface
+	discoveryClient  discovery.ServerResourcesInterface
+	crdClient        v1.CustomResourceDefinitionInterface
+	apiserviceClient apiregistrationv1.APIServiceInterface
+}
+
+func NewMigrator(kubeConfigPath string) (*migrator, error) {
+	restConfig, err := clientcmd.BuildConfigFromFlags("", kubeConfigPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build rest config: %w", err)
+	}
+	clientset, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build kubernetes clientset config: %w", err)
+	}
+	crd, err := crdclient.NewForConfig(restConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build crd clientset config: %w", err)
+	}
+	apiservice, err := apiserviceclient.NewForConfig(restConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build apiservice client config: %w", err)
+	}
+	dynamic, err := dynamic.NewForConfig(restConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build dynamic client config: %w", err)
+	}
+
+	return &migrator{
+		client:           dynamic,
+		discoveryClient:  clientset.Discovery(),
+		crdClient:        crd.ApiextensionsV1().CustomResourceDefinitions(),
+		apiserviceClient: apiservice.ApiregistrationV1().APIServices(),
+	}, nil
+}
+
+func (d *migrator) Start(ctx context.Context) (*MigrationResultList, error) {
+	return d.start(ctx)
+}
+
+func (d *migrator) start(ctx context.Context) (*MigrationResultList, error) {
+	schemas, err := d.findMigratableResources(ctx)
+	if err != nil {
+		return nil, err
+	}
+	results := &MigrationResultList{
+		Status: MigrationSuccess,
+	}
+	errorOccured := false
+	start := time.Now()
+	klog.Info("schema migration started")
+
+	wg := sync.WaitGroup{}
+	lock := sync.Mutex{}
+	wg.Add(len(schemas))
+	for _, sch := range schemas {
+		go func(wg *sync.WaitGroup, sch schema.GroupVersionResource) {
```

Comment:
How much of a performance benefit do we get by parallelizing this work? How many different workers are there running in parallel?

Reply:
Honestly I kept going back and forth on doing things concurrently here, figured this might help incase there are a lot of schemas, but at our size there is no appreciable performance gain, there would be a worker per schema group here. I wouldn't mind getting rid of it to reduce complexity

Reply:
Let's see how we can do without the complexity, then. If it turns into a problem, we'll have some understanding of _why_ and we can apply the right level of parallelism (a fixed number of workers, a worker per group, a worker per resource, etc.).

Reply:
Sounds good, removed for now